### PR TITLE
Add  Baleen -> JsonSchema overrides for type mapping

### PR DESCRIPTION
**Pull request by kdallmeyer-sr**
I have a need to do custom overrides when generating json schemas. I decided to add support for both from Baleen to json schema ~and from json schema to Baleen~.

~Utilized some Kotlin functional behaviors and reifed generics in order to make this work. (Some cool Kotlin stuff!) The end result is that a function can be created to generate the override and it will appropriately map the change.~

----- 
Edit 1/27/20

Re-wrote to reuse the pattern developed in XsdGenerator. I made it common in its own module and it will be used by JsonSchema generator and Xsd Generator.

### CODE REVIEW

File: baleen-jsonschema-generator/src/main/kotlin/com/shoprunner/baleen/jsonschema/v3/JsonSchemaGenerator.kt [hunk 1]
```diff
@@ -50,10 +50,15 @@ object JsonSchemaGenerator {
         }
     }
 
-    fun getJsonSchema(baleenType: BaleenType): JsonSchema {
+    fun getJsonSchema(baleenType: BaleenType, mappingOverrides: List<BaleenOverride> = emptyList()): JsonSchema {
```
Comment (HawaiianSpork) [inline baleen-jsonschema-generator/src/main/kotlin/com/shoprunner/baleen/jsonschema/v3/JsonSchemaGenerator.kt]: Rather then changing your code to handle overrides you can have `getJsonSchema` take a "type mapper".  You can then take your existing code and extract it into a function as the default "type mapper".  See how we do it in xsd generation as an example: https://github.com/ShopRunner/baleen/blob/master/baleen-xsd-generator/src/main/kotlin/com/shoprunner/baleen/xsd/XsdGenerator.kt#L132

Comment (kdallmeyer-sr) [reply baleen-jsonschema-generator/src/main/kotlin/com/shoprunner/baleen/jsonschema/v3/JsonSchemaGenerator.kt]: My one problem with this approach is that in order to override a single BaleenType you have to reconstruct the whole mapper. I've extracted out a Baleen mapper for `v4` and also created an OverrideMapper to provide point solutions.

**Comment by kdallmeyer-sr** [comment]
@HawaiianSpork can you look at https://github.com/ShopRunner/baleen/pull/19/commits/9d6b5abfa79ab1ad80f1688e6fbe93f67c61205a to see if looks like a reasonable middle ground?

File: baleen-jsonschema-generator/src/main/kotlin/com/shoprunner/baleen/jsonschema/v3/JsonSchemaGenerator.kt (hunk 1 above)
Comment (kdallmeyer-sr) [reply baleen-jsonschema-generator/src/main/kotlin/com/shoprunner/baleen/jsonschema/v3/JsonSchemaGenerator.kt]: @HawaiianSpork re-wrote it to follow the Xsd Generator pattern.

**Pull request merged**